### PR TITLE
test: Improvements to ArchiveFileComparator

### PR DESCRIPTION
**Pull request by jorsol**
Some improvements to the ArchiveFileComparator used in tests.

### CODE REVIEW

File: src/test/java/org/codehaus/plexus/archiver/zip/ArchiveFileComparator.java [hunk 1]
```diff
@@ -1,81 +1,76 @@
...
     {
-        for ( java.util.Enumeration en = file.getEntries(); en.hasMoreElements(); )
-        {
-            TarArchiveEntry ze = (TarArchiveEntry) en.nextElement();
-            if ( !ze.isDirectory() )
-            {
-                consumer.accept( ze );
-            }
-        }
+        tarEntries( file )
```
Comment (plamentotev) [inline src/test/java/org/codehaus/plexus/archiver/zip/ArchiveFileComparator.java:34]: Looks like other usages of `tarEntries` also skip the directories. Maybe we can rename it `tarFileEntries` or something like that and move the `!entry.isDirectory()` filter there to avoid code duplication.

File: src/test/java/org/codehaus/plexus/archiver/zip/ArchiveFileComparator.java [hunk 2]
```diff
@@ -84,119 +79,119 @@ private static Map getFileEntries( ZipFile file )
...
-        final InputStream is2 = file2.getInputStream( entry2 );
-        final byte[] bytes1 = IOUtil.toByteArray( is1 );
-        final byte[] bytes2 = IOUtil.toByteArray( is2 );
-        Assert.assertTrue( Arrays.equals( bytes1, bytes2 ) );
-        is1.close();
-        is2.close();
+        assertTrue( Math.abs( timeDelta ) <= 1000 );
+
+        try ( final InputStream is1 = file1.getInputStream( entry1 );
+                        final InputStream is2 = file2.getInputStream( entry2 ) )
```
Comment (plamentotev) [inline src/test/java/org/codehaus/plexus/archiver/zip/ArchiveFileComparator.java:127]: This line is not aligned with the one above.

**Review by plamentotev — COMMENTED** [review]
Great change. I've added a couple of comments.

File: src/test/java/org/codehaus/plexus/archiver/zip/ArchiveFileComparator.java [hunk 3]
```diff
@@ -84,119 +79,119 @@ private static Map getFileEntries( ZipFile file )
...
         }
-        Assert.assertTrue( map2.isEmpty() );
+        assertTrue( map2.isEmpty() );
     }
 
+    private static List<TarArchiveEntry> tarEntries( TarFile file )
+    {
+        try
+        {
+            return Collections.list( file.getEntries() )
```
[the remark this answers is not in the document]
Comment (jorsol) [reply src/test/java/org/codehaus/plexus/archiver/zip/ArchiveFileComparator.java:187]: I don't understand the question here, the getEntries() return an Enumeration, that's why to get the stream I need to convert it to a List first.

[the remark this answers is not in the document]
Comment (plamentotev) [reply src/test/java/org/codehaus/plexus/archiver/zip/ArchiveFileComparator.java]: What I meant is that `.collect(Collectors.toList())` returns `List` already. Why pass that to `Collectors.toList()` to convert it to `List`?

**Comment by plamentotev** [comment]
> Ok, I have done a complete rewrite of the class, now it should be cleaner and more consistent.
> 
> I have added AssertJ dependency to have better assert methods.

Thanks. I hope I'll have time to take a look in the next days. I guess AssertJ is ok for tests dependency.

File: src/test/java/org/codehaus/plexus/archiver/zip/ArchiveFileComparator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jorsol) [reply src/test/java/org/codehaus/plexus/archiver/zip/ArchiveFileComparator.java]: Oh, right, sorry, anyway the new refactor doesn't do a `Collectors.toList()` anymore.

Essentially to be able to get a stream the easiest way is to convert the enumeration to a List first, (an Enumeration does not support Stream API OOTB).

Why a stream is needed? because is the easiest way to filter, transform, and collect the elements, the `TarFile.getEntries()` returns an `ArchiveEntries`, and to use the consumer a `TarArchiveEntry` is expected so a cast is needed (the current API is a bit weird).

In other words `file.getEntries()` returns an `Enumeration` and to convert an enumeration to a `List` I use the `Collections.list()`. (Collections != Collectors)

File: src/test/java/org/codehaus/plexus/archiver/zip/ArchiveFileComparator.java [hunk 4]
```diff
@@ -1,202 +1,169 @@
...
     }
 
-    private static Map getFileEntries( ZipFile file )
-        throws IOException
+    private static void assertTarEquals( TarFile file1, TarArchiveEntry entry1,
+                                         TarFile file2, TarArchiveEntry entry2 )
     {
-        final Map map = new HashMap();
-        for ( java.util.Enumeration en = file.getEntries(); en.hasMoreElements(); )
+        assertThat( entry1.isDirectory() ).isEqualTo( entry2.isDirectory() ).isFalse();
```
Comment (plamentotev) [inline src/test/java/org/codehaus/plexus/archiver/zip/ArchiveFileComparator.java:76]: Why asserting that the entry is not directory? I think the current just checks if both are equal.

File: src/test/java/org/codehaus/plexus/archiver/zip/ArchiveFileComparator.java [hunk 5]
```diff
@@ -1,202 +1,169 @@
...
     {
-        final Map map1 = getFileEntries( file1 );
-        final Map map2 = getFileEntries( file2 );
-        for ( Object o : map1.keySet() )
-        {
-            final String name1 = (String) o;
-            final String name2 = prefix == null ? name1 : ( prefix + name1 );
-            ZipArchiveEntry ze1 = (ZipArchiveEntry) map1.get( name1 );
-            ZipArchiveEntry ze2 = (ZipArchiveEntry) map2.remove( name2 );
-            Assert.assertNotNull( "Missing " + name1 + "in file 2", ze2 );
```
Comment (plamentotev) [inline src/test/java/org/codehaus/plexus/archiver/zip/ArchiveFileComparator.java:196]: I would keep the message as it would make it quicker to understand why the test is failing.

File: src/test/java/org/codehaus/plexus/archiver/zip/ArchiveFileComparator.java [hunk 6]
```diff
@@ -1,202 +1,169 @@
...
+              InputStream is2 = file2.getInputStream( entry2 ); )
         {
-            ZipArchiveEntry ze = (ZipArchiveEntry) en.nextElement();
-            if ( ze.isDirectory() )
-            {
-                continue;
-            }
-            if ( map.put( ze.getName(), ze ) != null )
-            {
-                Assert.fail( "Multiple archive file entries named " + ze.getName() + " found." );
```
Comment (plamentotev) [inline src/test/java/org/codehaus/plexus/archiver/zip/ArchiveFileComparator.java:78]: Maybe I'm missing something, but to me it looks like the new version would not verify if that the same entry exists only once in the archive.

Comment (jorsol) [reply src/test/java/org/codehaus/plexus/archiver/zip/ArchiveFileComparator.java:78]: There is an implicit check with the Collector.toMap as it uses a uniqKeysMapAccumulator that throws an exception on a duplicate key.

**Comment by plamentotev** [comment]
Great work, thanks.
